### PR TITLE
feat(aria): add aria labeling to the trigger

### LINKED ISSUE #77: Label content is not accessible
## Describe the bug

The label slot is simply standard content and does not read back as a label for the interactive element. 

## Expected behavior

Content should be read back as a label for interactive content. Consider https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-labelledby

## Additional context

This update should be reviewed with the auro-select element as well. 

AC/Exit Criteria:
User should be able to define custom aria-label.  If no custom label is defined and label slot exists then use that as aria label.

### PR DESCRIPTION
**Pull request by jason-capsule42**
# Alaska Airlines Pull Request

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

**Fixes:** #77

## Summary:

This change is meant to get a valid aria labeling rule on the trigger element.

If a custom `aria-label` is applied at the root that will get applied to the trigger elem.
Else if a custom `aria-labeledby` is applied to the root that will get applied.
Else if the `label` slot has content that will get applied.

After reviewing the implementation of this with @geoffrich it was understood that *all* instances of `auro-dropdown` need to have `aria-label` or `aria-labelledby` on the trigger or the voiceover experience will not meet expectations. An update has been made to the demo page explaining this and all apiExamples have been updated to be compliant with this expectations.

## Type of change:

Please delete options that are not relevant.

- [x] New capability
- [ ] Revision of an existing capability
- [ ] Infrastructure change (automation, etc.)
- [ ] Other (please elaborate)

## Checklist:

- [x] My update follows the CONTRIBUTING guidelines of this project
- [x] I have performed a self-review of my own update

**By submitting this Pull Request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

_Pull Requests will be evaluated by their quality of update and whether it is consistent with the goals and values of this project. Any submission is to be considered a conversation between the submitter and the maintainers of this project and may require changes to your submission._

**Thank you for your submission!**<br>
-- Auro Design System Team

### CODE REVIEW

File: src/auro-dropdown.js [hunk 1]
```diff
@@ -94,7 +94,32 @@ class AuroDropdown extends LitElement {
...
+   * @private
+   * @returns {void} Updates the trigger `aria-label` or `aria-labeledby`.
+   */
+  processAriaLabel() {
+    this.trigger = this.shadowRoot.querySelector(`#trigger`);
+
+    if (this.getAttribute('aria-label')) {
+      // If custom aria-label declared at root apply it to the trigger
+      this.trigger.setAttribute('aria-label', this.getAttribute('aria-label'));
+    } else if (this.getAttribute('aria-labeledby')) {
```
Comment (geoffrich) [inline src/auro-dropdown.js]: This attribute is misspelled - it's [aria-labelledby](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-labelledby)

File: src/auro-dropdown.js [hunk 2]
```diff
@@ -94,7 +94,32 @@ class AuroDropdown extends LitElement {
...
+  /**
+   * @private
+   * @returns {void} Updates the trigger `aria-label` or `aria-labeledby`.
+   */
+  processAriaLabel() {
+    this.trigger = this.shadowRoot.querySelector(`#trigger`);
+
+    if (this.getAttribute('aria-label')) {
+      // If custom aria-label declared at root apply it to the trigger
+      this.trigger.setAttribute('aria-label', this.getAttribute('aria-label'));
```
Comment (geoffrich) [inline src/auro-dropdown.js]: I don't like imperatively setting the attribute like this. Can we do it in the template instead?

File: src/auro-dropdown.js [hunk 3]
```diff
@@ -94,7 +94,32 @@ class AuroDropdown extends LitElement {
...
+    } else if (this.getAttribute('aria-labeledby')) {
+      // Else if custom aria-labeledby declared at root apply it to trigger
+      this.trigger.setAttribute('aria-labeledby', this.getAttribute('aria-labeledby'));
```
Comment (geoffrich) [inline src/auro-dropdown.js]: I'm not sure if this will be valid. We should test this behavior with a screen reader. My concern is that `<auro-dropdown aria-labelledby="foo">` will refer to an element with ID "foo" outside of this component's shadow root, and using `aria-labelledby="foo"` inside the shadow root won't be able to reference that element.

Comment (jason-capsule42) [reply src/auro-dropdown.js]: This works correctly for me using VoiceOver on my mac. If you run the localhost demo page that last example in the demo renders with this method. @geoffrich please let me know if your mileage is different with this in the windows environment! I'd appreciate the feedback after trying out that coded example.

File: src/auro-dropdown.js (hunk 2 above)
Comment (jason-capsule42) [reply src/auro-dropdown.js]: Can you expand on a reason to move this more than "don't like"?

We can certainly move it to the template but I think that code is likely heavier and requires a little more mental leaps to account for the logic of when to use `label` vs `labelledby` and not end up with both attributes on the trigger.

Comment (geoffrich) [reply src/auro-dropdown.js]: Sorry, I can clarify - I'm fine with keeping the if/else logic here. But I don't think we should be imperatively setting attributes inside the component with `this.trigger.setAttribute`. In general, if we're modifying the component's template, we should do it in the render method. Doing it imperatively in a different method can make it difficult to track down where it's happening. Instead, we should set a state variable and then apply it in the render method. So I'd prefer something like this... 

```
else if (this.getAttribute('aria-labelledby')) {
      // Else if custom aria-labelledby declared at root apply it to trigger
      this.labelledBy = this.getAttribute('aria-labelledby');
    }
// ...then in the render method

<div id="trigger" class="trigger" data-trigger-placement="${this.placement}" aria-labelledby=${this.labelledBy}>
```

I think doing it this way is also more performant, since Lit can optimize the state updates, though I'm not 100% certain.

Initializing `this.labelledBy` to undefined should make sure the attribute is not set by default (IIRC).

File: src/auro-dropdown.js (hunk 3 above)
Comment (geoffrich) [reply src/auro-dropdown.js]: This only worked once I added `role="button"` to the trigger. Otherwise, NVDA wouldn't announce the labels (presumably because the trigger wasn't exposed as an interactive element). Can you make that change?

File: src/auro-dropdown.js (hunk 2 above)
Comment (jason-capsule42) [reply src/auro-dropdown.js]: @geoffrich This change was achieved using `ifDefined` rules in the template along with the change to reference properties rather than apply the attributes directly with JavaScript.

**Comment by jason-capsule42** [comment]
@blackfalcon I am done making updates to this PR based on the conversation with @geoffrich. I think this is ready for a final review from you.

File: demo/demo.md [hunk 4]
```diff
@@ -32,10 +32,20 @@ The dropdown can be closed with the following actions:
     - `spacebar`
     - `esc`
 
+## Accessibility requirements
+
+All instances of `auro-dropdown` must have a valid `aria-label` or `aria-labelledby` attribute. There are three options available to meet this requirement. Only one of these options should be implemented.
```
Comment (geoffrich) [inline demo/demo.md]: This is incorrect - the label slot is also an option, and doesn't require the dev to set aria-label or aria-labelledby themselves. Aria attributes should only be used when absolutely necessary. Can we re-word this to suggest the label slot as the primary method of labelling the dropdown? 

I'd also recommend updating the examples to use the label slot instead of aria-label.

File: demo/demo.md [hunk 5]
```diff
@@ -32,10 +32,20 @@ The dropdown can be closed with the following actions:
...
 
+## Accessibility requirements
+
+All instances of `auro-dropdown` must have a valid `aria-label` or `aria-labelledby` attribute. There are three options available to meet this requirement. Only one of these options should be implemented.
+
+- Add the `aria-label` attribute to the dropdown root (e.g. `<auro-dropdown aria-label="value">`)
+- Add the `aria-labelledby` attribute to the dropdown root (e.g. `<auro-dropdown aria-labelledby="idOfElement">`)
+- Include content in the `label` slot. When doing so, the `aria-labelledby` attribute will automatically reference the label slot content.
+
+Failing to include one of the above options will result in voiceover utilities being unable to correctly identify the dropdown to the user.
```
Comment (geoffrich) [inline demo/demo.md]:
```suggestion
Failing to include one of the above options will prevent assistive technology from correctly identifying the dropdown to the user.

```

File: docs/partials/demo.md [hunk 6]
```diff
@@ -27,6 +27,16 @@ The dropdown can be closed with the following actions:
     - `spacebar`
     - `esc`
 
+## Accessibility requirements
+
+All instances of `auro-dropdown` must have a valid `aria-label` or `aria-labelledby` attribute. There are three options available to meet this requirement. Only one of these options should be implemented.
```
Comment (geoffrich) [inline docs/partials/demo.md]: This is a duplicate of the content in `demo.md` - we should only have one.

File: src/auro-dropdown.js [hunk 7]
```diff
@@ -47,6 +48,7 @@ class AuroDropdown extends LitElement {
     this.inset = false;
     this.rounded = false;
     this.toggle = false;
+    // this.ariaLabel = 'test label from prop';
```
Comment (geoffrich) [inline src/auro-dropdown.js]: Remove commented out code

File: src/auro-dropdown.js [hunk 8]
```diff
@@ -94,7 +106,32 @@ class AuroDropdown extends LitElement {
     this.dropdownWidth = this.getBoundingClientRect().width;
   }
 
+  /**
+   * @private
+   * @returns {void} Updates the trigger `aria-label` or `aria-labelledby`.
+   */
+  processAriaLabel() {
+    this.trigger = this.shadowRoot.querySelector(`#trigger`);
```
Comment (geoffrich) [inline src/auro-dropdown.js]: This is a duplicate of line 100 - can we remove this line and this function call after the trigger is set?

File: src/auro-dropdown.js [hunk 9]
```diff
@@ -219,13 +256,19 @@ class AuroDropdown extends LitElement {
       <div id="popover" class="popover" aria-live="polite" style=${`min-width: ${this.dropdownWidth}px;`}>
         <slot role="tooltip"></slot>
       </div>
-      <div id="trigger" class="trigger" data-trigger-placement="${this.placement}">
+      <div
+        id="trigger"
+        class="trigger"
+        role="button"
+        aria-label="${ifDefined(this.ariaLabel ? this.ariaLabel : undefined)}"
+        aria-labelledby="${ifDefined(this.ariaLabelledby ? this.ariaLabelledby : undefined)}"
```
Comment (geoffrich) [inline src/auro-dropdown.js]:
```suggestion
        aria-label="${ifDefined(this.ariaLabel)}"
        aria-labelledby="${ifDefined(this.ariaLabelledby)}"
```

I _think_ this should work instead - asumming ariaLabel and ariaLabelledby defaults to undefined

File: src/auro-dropdown.js [hunk 10]
```diff
@@ -219,13 +256,19 @@ class AuroDropdown extends LitElement {
...
+        data-trigger-placement="${this.placement}">
         <div class="triggerContentWrapper">
-          <div class="label">
+          <div class="label" id="triggerLabel">
             <slot name="label"></slot>
           </div>
           <div class="triggerContent" chevron=${this.chevron}>
-            <slot name="trigger"></slot>
+            <slot name="trigger"></slot>${this.ariaLabel}
```
Comment (geoffrich) [inline src/auro-dropdown.js]: Why was this added?

File: demo/demo.md (hunk 4 above)
Comment (jason-capsule42) [reply demo/demo.md]: @geoffrich The way I wrote this is meant to say that one way or another you must get an `aria-label` or `aria-labelledby` attribute on the trigger. Using the slot is just another means of making that end goal happen.

Regarding if the slot should be our default example or not - I will bring this up during our standup discussion today with @leeejune. I'd like design to weigh in here. I had understood the label slot to be optional but not the default implementation choice.

File: demo/demo.md (hunk 5 above)
Comment (jason-capsule42) [reply demo/demo.md]: I like the rewording here, @geoffrich. I was struggling trying to find the right words. Great suggestion. Thank you.

File: demo/demo.md (hunk 4 above)
Comment (geoffrich) [reply demo/demo.md]: Ah, gotcha - we can go with whatever design says on that front. I still think this is worded confusingly; the user of the component shouldn't care that the label slot sets aria-labelledby behind the scenes.

File: docs/partials/demo.md (hunk 6 above)
Comment (jason-capsule42) [reply docs/partials/demo.md]: This is the template that our compiled ./demo/demo.md builds from. This is where the change is made manually.

File: src/auro-dropdown.js (hunk 8 above)
Comment (jason-capsule42) [reply src/auro-dropdown.js]: I may be misunderstanding here due to the line numbers changing after making updates from other commits, but I don't see where this is already declared elsewhere in the method.

File: src/auro-dropdown.js (hunk 9 above)
Comment (jason-capsule42) [reply src/auro-dropdown.js]: You were right, the shorthand version does work. Thanks, @geoffrich.

File: src/auro-dropdown.js (hunk 8 above)
Comment (geoffrich) [reply src/auro-dropdown.js]: We also set the trigger in the `firstUpdated` method, where this method is called from. We should only set it once

File: demo/demo.md (hunk 4 above)
Comment (jason-capsule42) [reply demo/demo.md]: I agree the wording could improve. I put a tweak in that may be less confusing by saying less. :)

Let me know what you think, @geoffrich.

Comment (geoffrich) [reply demo/demo.md]: Better, but I think we can make it even simpler. How about "All instances of `auro-dropdown` must have a valid label." for the first sentence? Then the rest is good

**Review by blackfalcon — CHANGES_REQUESTED** [review]
This branch and PR should be updated to be a PATCH release. The issue was reported as a bug and this update resolves an issue related to a broken experience, not add a new feature to the element.

**Comment by blackfalcon** [comment]
@geoffrich @jason-capsule42 

I am curious as to this implementation. When I made this issue and linked to the `aria-labeledby` recommendation, the intent that I had was that we would use this to bind the content from the string in the `label` slot to this aria rule. 

The error that I encountered was that since focusing on the trigger totally bypasses the label, it is never read back to the user. 

The following is the update I had in mind. 

```js
  render() {
    return html`
      <div id="popover" class="popover" aria-live="polite" style=${`min-width: ${this.dropdownWidth}px;`}>
        <slot role="tooltip"></slot>
      </div>
      <div
        id="trigger"
        class="trigger"
        role="button"
        data-trigger-placement="${this.placement}">
        <div class="triggerContentWrapper">
          <label class="label" id="triggerLabel">
            <slot name="label"></slot>
          </label>
          <div class="triggerContent" chevron=${this.chevron} aria-labeledby="triggerLabel">
            <slot name="trigger"></slot>
          </div>
        </div>
        ${this.chevron ? html`
          <div id="showStateIcon">
            <auro-icon
              category="interface"
              name="chevron-down"
              customColor>
            </auro-icon>
          </div>
        ` : undefined}
      </div>
      <div class="helpText">
        <slot name="helpText"></slot>
      </div>
    `;
  }
}
```

With this update, the content in the `label` slot will take priority over the use of `aria-label` at the root. But, using `aria-label` at the root address the use case for when the `label` slot is not used. 

Where this does fall short is using `aria-labeledby` to content outside the scope of the element. But I am not sure that is a use case we need to support. Either the `label` slot is used or `aria-label` on the root is used instead of label content. 

What is the advantage of the current solution versus what I am proposing?

File: package.json [hunk 11]
```diff
@@ -142,6 +142,7 @@
     "build:dev:assets": "npm-run-all build:sass:demo build:sass:component postCss:component sass:render",
     "build:markdownDocs": "node scripts/generateDocs.js",
     "build:sass": "npm-run-all build:sass:demo build:sass:component postCss:component sass:render",
+    "build:md": "npm run build:markdownDocs",
```
Comment (jason-capsule42) [inline package.json]: This change should not be part of this PR. If we want to make a shorthand of `build:markdownDocs` (which I am in favor of) that change needs to be in WC-Generator.

**Comment by jason-capsule42** [comment]
@blackfalcon The update you made has two notable impacts:

1. The `aria-labelledby` attribute is left on the trigger content even when it's not going to be in use. This doesn't seem to cause a problem for VoiceOver on my mac. @geoffrich would you be able to check using NVDA?
1. It's no longer possible for a custom `aria-labelledby`. I'm not sure why we would choose to support a custom `aria-label` but not `aria-labelledby`.

Otherwise, the change seems fine.

**Comment by blackfalcon** [comment]
What is the scenario for a custom labeledby? 

The only thing I can think of is attaching this to something else in the DOM outside the scope of the element. If that is the case, that has to happen on the parent container of the element, aria does not pierce the shadow DOM.

File: demo/demo.md [hunk 12]
```diff
@@ -89,17 +127,22 @@ Failing to include one of the above options will prevent assistive technology fr
 ```
 
 </auro-accordion>
+
+## Common use using the label content slot
```
Comment (jason-capsule42) [inline demo/demo.md:131]: "Common use with the label content slot" is consistent with the headers on other examples
```suggestion
## Common use with the label content slot
```

File: demo/demo.md [hunk 13]
```diff
@@ -138,3 +181,18 @@ Failing to include one of the above options will prevent assistive technology fr
...
+## Supported standard and accessible interactions
+
+The dropdown can be opened with the following actions:
+
+1. Clicking/tapping on the trigger will open the dropdown UI.
+1. Tabbing to the trigger and using the `enter` or `spacebar` keys will open the dropdown UI.
+
+The dropdown can be closed with the following actions:
+
+1. Clicking anywhere in the view outside of the dropdown.
```
Comment (jason-capsule42) [inline demo/demo.md:194]: How about `document` or `window` instead of `view` which I don't think is a correct term here?

File: demo/demo.md [hunk 14]
```diff
@@ -138,3 +181,18 @@ Failing to include one of the above options will prevent assistive technology fr
...
+1. When the `toggle` attribute is added, clicking on the trigger.
+1. Programmatically via another control in the UI calling the `hide()` method (see api).
```
Comment (jason-capsule42) [inline demo/demo.md:196]: Doesn't have to be from a UI control.
```suggestion
1. Programmatically via calling the `hide()` method (see api).
```

File: demo/demo.md [hunk 15]
```diff
@@ -138,3 +181,18 @@ Failing to include one of the above options will prevent assistive technology fr
...
+1. Selecting an option within the dropdown by either keyboard or click/tap.
```
Comment (jason-capsule42) [inline demo/demo.md]: There is no code in dropdown itself that does this. For instance in the case of `auro-select` making a selection from the menu calls `dropdown.hide()`.

We should remove this line.

File: demo/demo.md [hunk 16]
```diff
@@ -128,3 +181,18 @@ The dropdown can be closed with the following actions:
...
+## Supported standard and accessible interactions
+
+The dropdown can be opened with the following actions:
+
+1. Clicking/tapping on the trigger will open the dropdown UI.
+1. Tabbing to the trigger and using the `enter` or `spacebar` keys will open the dropdown UI.
+
+The dropdown can be closed with the following actions:
+
+1. Clicking anywhere in the view outside of the dropdown.
```
Comment (blackfalcon) [inline demo/demo.md:194]:
```suggestion
1. Clicking anywhere in the wndow outside of the dropdown.
```

File: demo/demo.md [hunk 17]
```diff
@@ -128,3 +181,18 @@ The dropdown can be closed with the following actions:
...
+1. When the `toggle` attribute is added, clicking on the trigger.
+1. Programmatically via another control in the UI calling the `hide()` method (see api).
+1. Selecting an option within the dropdown by either keyboard or click/tap.
```
Comment (blackfalcon) [inline demo/demo.md]:
```suggestion
```

**Review by blackfalcon — DISMISSED** [review]
